### PR TITLE
support reserved node

### PR DESCRIPTION
**Pull request by qiankunli**
for example, my k8s cluster has 10 nodes (10 gpu per node), 5 for training, 5 for serving, I want use volcano schedule tfjob/pytorchjob/vcjob on training nodes. use default-scheduler  schedule online pod on serving nodes.

it is work by setting nodeSelector or nodeAffinity to pod,  but it is not properly when considering queue.

if there are  two queue: queue1(weight=1) and queue2 (weight=1),  it is overused if queue1 used 40 gpu, but it's weight  resource is 50, so if queue2 want use 20 gpu, reclaimable mechanism will not work.

so it is necessary to limit volcano scheduler that  it is can only work on training nodes, not all nodes in cluster.  

so I add `work-node-label` args for volcano scheduler, in following example,  volcano can only work on the node which has label `node-role:training` or `node-role:dev`. you can use any label name you like.

```yaml
kind: Deployment
apiVersion: apps/v1
metadata:
  name: volcano-scheduler
  namespace: xdl-system
  labels:
    app: volcano-scheduler
spec:
  replicas: 1
  selector:
    matchLabels:
      app: volcano-scheduler
  template:
    metadata:
      labels:
        app: volcano-scheduler
    spec:
      serviceAccount: volcano-scheduler
      containers:
        - name: volcano-scheduler
          image: xx/vc-scheduler:v1.4.0-node
          args:
            - --work-node-label=node-role:training
            - --work-node-label=node-role:dev
            - --logtostderr
            - --scheduler-conf=/volcano.scheduler/volcano-scheduler.conf
            - -v=5
            - 2>&1
          imagePullPolicy: "IfNotPresent"
          volumeMounts:
            - name: scheduler-config
              mountPath: /volcano.scheduler
      volumes:
        - name: scheduler-config
          configMap:
            name: volcano-scheduler-configmap
```

### CODE REVIEW

**Comment by Thor-wl** [comment]
@qiankunli Pls `go fmt xxx`

File: cmd/scheduler/app/options/options.go [hunk 1]
```diff
@@ -106,6 +108,7 @@ func (s *ServerOption) AddFlags(fs *pflag.FlagSet) {
 	fs.Int32Var(&s.PercentageOfNodesToFind, "percentage-nodes-to-find", defaultPercentageOfNodesToFind, "The percentage of nodes to find and score, if <=0 will be calcuated based on the cluster size")
 
 	fs.StringVar(&s.PluginsDir, "plugins-dir", defaultPluginsDir, "vc-scheduler will load custom plugins which are in this directory")
+	fs.StringSliceVar(&s.WorkNodeLabels, "work-node-label", nil, "volcano only work with the labeled node, like: --work-node-label=volcano.sh/role:train --work-node-label=volcano.sh/role:serving")
```
Comment (k82cn) [inline cmd/scheduler/app/options/options.go:111]: Prefer to check those labels in `CheckOptionOrDie` :)

**Comment by Thor-wl** [comment]
> if there are two queue: queue1(weight=1) and queue2 (weight=1), it is overused if queue1 used 40 gpu, but it's weight resource is 50, so if queue2 want use 20 gpu, reclaimable mechanism will not work.

Sorry for not clear why queue1 is overused if it used 40 gpu for it's weight is 1.

**Comment by Thor-wl** [comment]
I agree with the design to select nodes which volcnao can charge with node label, but it's not a flexible implementation way to specify `work-node` as arguments in start command. Suggent to consider set them in the configuration of scheduler.

**Comment by Thor-wl** [comment]
BTW, this feature is import and useful. Suggest to add a doc for that.

**Comment by qiankunli** [comment]
> BTW, this feature is import and useful. Suggest to add a doc for that.

ok, I add a doc    https://github.com/qiankunli/volcano/blob/feature/support_reserved_node/docs/design/node-selector.md

**Comment by Thor-wl** [comment]
> > BTW, this feature is import and useful. Suggest to add a doc for that.
> 
> ok, I add a doc https://github.com/qiankunli/volcano/blob/feature/support_reserved_node/docs/design/node-selector.md

OK. Suggest to submit a PR and add the doc to docs/design

**Comment by shinytang6** [comment]
> /cc @allReviewers This feature is important and meaningful. I plan to take it into v1.5. Pls give your review comments. THX.

It's better to set up some sig or wg(working group) under volcano org, then you can @ all members to review :) @Thor-wl

File: pkg/scheduler/cache/cache.go [hunk 2]
```diff
@@ -45,7 +45,11 @@ import (
 	"k8s.io/klog"
 	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
 	volumescheduling "k8s.io/kubernetes/pkg/controller/volume/scheduling"
-
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
```
Comment (shinytang6) [inline pkg/scheduler/cache/cache.go:52]: please help make the ci happy first

**Review by shinytang6 — CHANGES_REQUESTED** [review]

**Review by jasonliu747 — COMMENTED** [review]
/hold

**Comment by Thor-wl** [comment]
> 

Good idea! This event will take into schedule in Q1 2022. Need all the contributors' advices.

**Comment by shinytang6** [comment]
ls this ready for review? /cc @qiankunli

File: pkg/scheduler/cache/cache.go [hunk 3]
```diff
@@ -364,22 +366,40 @@ func newSchedulerCache(config *rest.Config, schedulerName string, defaultQueue s
...
 
 		NodeList: []string{},
 	}
+	if len(nodeSelector) > 0 {
+		for _, nodeSelectorLabel := range nodeSelector {
+			nodeSelectorLabelLen := len(nodeSelectorLabel)
+			if nodeSelectorLabelLen <= 0 {
+				continue
+			}
+			//sc.nodeSelectorLabels[nodeSelectorLabel] = ""
```
Comment (shinytang6) [inline pkg/scheduler/cache/cache.go:390]: pls remove the unused comments

File: cmd/scheduler/app/options/options.go [hunk 4]
```diff
@@ -66,6 +66,8 @@ type ServerOption struct {
 	MinNodesToFind             int32
 	MinPercentageOfNodesToFind int32
 	PercentageOfNodesToFind    int32
+
+	NodeSelector []string
```
Comment (shinytang6) [inline cmd/scheduler/app/options/options.go:70]: NodeSelector => NodeSelectors

**Review by shinytang6 — CHANGES_REQUESTED** [review]

**Review by shinytang6 — COMMENTED** [review]
lgtm overall :)

**Pull request merged**